Dashboard v2: User Profile - Account deletion

brunobasto opened the pull request:
Fixes https://linear.app/a8c/issue/DOTDASH-369/implement-profile-account-deletion-section

## Proposed Changes

This is a follow-up on https://github.com/Automattic/wp-calypso/pull/105491

Continuing the work on the user profile page, this PR adds the account deletion card proposed by the new designs:

<img width="693" height="134" alt="Screenshot 2025-09-03 at 14 57 33" src="https://github.com/user-attachments/assets/07c1dd1e-95c9-4b88-b844-780faca9a2f3" />

<img width="523" height="537" alt="Screenshot 2025-09-03 at 14 57 42" src="https://github.com/user-attachments/assets/56ff76ae-e388-450d-95b6-08ab5209ae16" />

<img width="660" height="275" alt="Screenshot 2025-09-03 at 14 57 49" src="https://github.com/user-attachments/assets/4ca77a20-d60b-43f0-b1d6-60cdb17b5834" />

### The "Are you sure?" modal

I couldn't find the respective designs for this modal, so I asked on `#hosting-dashboard-pair-programming` and @taipeicoder suggested porting over the functionality, using new components when possible. I went with `<ActionList>` and `<ActionItem>`, which felt appropriate for this use case IMO.

#### The links

I tried to map the old links to their corresponding mapping on v2, but I couldn't find one for `http://calypso.localhost:3000/start?ref=me-account-close` so I kept the old link for now.

After confirming you want to delete your account, following the same pattern of v1, you get redirected to `/me/account/closed` - which is a route we don't have implemented yet, and I don't see it yet on the new designs. I've created [DOTDASH-381](https://linear.app/a8c/issue/DOTDASH-381/implement-the-meaccountclosed-page) to track this work.

## Why are these changes being made?

To implement the proposed designs for the new account profile page.

## Testing Instructions

⚠️ Do this with a test account to make sure you don't delete your personal account.

1. Navigate to http://calypso.localhost:3000/v2/me/profile
2. Scroll to the bottom of the page
3. Assert that you see the account deletion section:

<img width="693" height="134" alt="Screenshot 2025-09-03 at 14 57 33" src="https://github.com/user-attachments/assets/43ca7311-066c-4d11-b6f1-90bbb4453d18" />

4. Click "Delete account"
5. Assert that you see the "Are you sure" modal:

<img width="523" height="537" alt="Screenshot 2025-09-03 at 14 57 42" src="https://github.com/user-attachments/assets/66626b98-74fc-4084-8298-49cf272ab459" />

6. Click "Continue"
7. Assert that you see the confirmation modal:

<img width="660" height="275" alt="Screenshot 2025-09-03 at 14 57 49" src="https://github.com/user-attachments/assets/6fae4f0e-9798-4e19-a038-2636cc5a88b6" />

8. Type you user name and click "Delete account"
8. Assert that your account is deleted and that you are redirected to `/me/account/closed`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you tested accessibility for your changes? Ensure the feature remains usable with various user agents (e.g., browsers), interfaces (e.g., keyboard navigation), and assistive technologies (e.g., screen readers) (PCYsg-S3g-p2).
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
    - [ ] For UI changes, have we tested the change in various languages (for example, ES, PT, FR, or DE)? The length of text and words vary significantly between languages. 
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

brunobasto committed:
Add api queries to allow closing and restoring an account

brunobasto committed:
Add account deletion section to profile page

brunobasto committed:
Make confirmation modal show alternatives to account deletion - like v1 used to do

brunobasto committed:
Add some tests to assert modal behavior

arthur791004 commented on client/dashboard/me/profile-deletion/index.tsx:
```diff
@@ -0,0 +1,84 @@
+				</CardBody>
+			</Card>
+
+			<AccountDeletionConfirmModal
```
We can simply render the modal only when the `isOpen` is true. It's also the example usage provided by the modal component 🙂

brunobasto replied:
Right! Much better. Done in https://github.com/Automattic/wp-calypso/pull/105544/commits/7d1f265fe8873e3a5284a9dd8ad007644d2de0fe

arthur791004 commented on client/dashboard/me/profile-deletion/index.tsx:
```diff
@@ -0,0 +1,84 @@
+	const mutation = useMutation( closeAccountMutation() );
+
+	// Ensure we have a username before rendering
+	if ( ! user?.username ) {
```
Is there any issue about this? If the user is not available, the `AuthProvider` won't render anything. As a result, I think this situation won't happen 🤔

brunobasto replied:
Good point. That wasn't needed. Resolved here https://github.com/Automattic/wp-calypso/pull/105544/commits/e0286f1d246173b144f38471795fd4b3d08a15ce

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+			onRequestClose={ handleClose }
+		>
+			<VStack spacing={ 4 }>
+				{ showAlternatives ? (
```
It would be better to separate these two cases as follows:

```js
if ( showAlternatives ) {
  return (
    <Modal>...</Modal>
  );
}

return (
  <Modal>...</Modal>
);
```

Alternatively, we could create a function to render the content and avoid duplication. However, I assume the two cases differ enough that this might not be necessary.

```js
<Modal
	title={ showAlternatives ? __( 'Are you sure?' ) : __( 'Confirm account deletion' ) }
	onRequestClose={ handleClose }
>
	<VStack spacing={ 4 }>
		{ renderContent() }
	</VStack>
</Modal>
```

brunobasto replied:
I like your first suggestion. Implemented in https://github.com/Automattic/wp-calypso/pull/105544/commits/d9aff6992a6a6a89404f81fa5f3c8489d1cbcd8e

brunobasto replied:
I'm also considering splitting that into two separate components..

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+	// Reset state when modal closes
+	const handleClose = () => {
+		setConfirmText( '' );
+		setShowAlternatives( true );
```
If we render the modal only when the `isOpen` is true, we don't need to reset the state 🙂

brunobasto replied:
Very true. Done in https://github.com/Automattic/wp-calypso/pull/105544/commits/50122bf019e1090d398b535b631c0e53be5b1717

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+											) }
+											<Button
+												variant="tertiary"
+												icon={ chevronRight }
```
We should use `isRTL()` to determine the direction of the icon to use

brunobasto replied:
Good catch. Added here: https://github.com/Automattic/wp-calypso/pull/105544/commits/883886f657c57d42a6745a904f65041b7d0ff108

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+													target="_blank"
+													rel="noopener noreferrer"
+													aria-label={ __( 'Get help' ) }
+													style={ { minWidth: 'auto' } }
```
Would it be better to set the spacing of the `HStack` to 0 instead of changing the minWidth here as the hover state looks unbalanced 🤔

brunobasto replied:
I removed the minWidth styling, as it was unnecessary: https://github.com/Automattic/wp-calypso/pull/105544/commits/3b9650fca88f38c9275a6bde6a93b6510c98d209

> he hover state looks unbalanced 🤔

I'm not 100% sure what you meant with "unbalanced"but I think you're referring to the discontinuity when you hover the mouse horizontally and there's a gap where there's no hover state?

I've applied you suggestion in https://github.com/Automattic/wp-calypso/pull/105544/commits/a27f24bae6b719bd00a0057b4638ba80f8a4e37f and I agree the continuity of the hover state feels better. I initially added the spacing because I didn't want the two buttons to be too close. But looking at it now I don't think it's an issue. Looks good to me.

arthur791004 replied:
I mean the vertical padding is different from the horizontal padding 😄

brunobasto replied:
Ah I see!

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+								'Please type your username in the field below to confirm. Your account will then be gone forever.'
+							) }
+						</Text>
+						<TextControl
```
We'd prefer to use `DataForm` for anything like this to keep the consistency

brunobasto replied:
Added DataForm here: https://github.com/Automattic/wp-calypso/pull/105544/commits/ed8026e9db5f530ca11d1fdf982e2111379b3f3c

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,180 @@
+									actions={
+										<HStack spacing={ 2 } expanded={ false }>
+											{ option.supportLink && (
+												<Button
```
It would be better to use `RouterLinkButton` to get the benefits of preload, etc.

brunobasto replied:
I've applied `RouterLinkButton` in https://github.com/Automattic/wp-calypso/pull/105544/commits/e9574eda0dcc4f05814e5d38e6751a436534fc44 but only to the internal links, as I'm assuming it shouldn't be used for external links like the support links.

I've also removed `target=_blank` from those links as I assume it doesn't make sense to preload things that will render in another window - preloader is probably smart enough to don't do that - so I removed it to get the benefits of preloading.

arthur791004 replied:
Make sense to me 👍

brunobasto committed:
Does not render modal component when not needed

brunobasto committed:
Remove unnecessary username check

brunobasto committed:
No need to reset state anymore, since modal is unmounted when closed

brunobasto committed:
Remove unnecessary minWidth styling

brunobasto committed:
Use spacing=0 for better continuity of the hover state

brunobasto committed:
Use RouterLinkButton for internal links

This way we get the benefit of preload. I've also removed the target blank since preload wouldn't be a benefit if the link opens on a new window.

brunobasto committed:
Fix missing spot when renaming `href` to `to`

brunobasto committed:
Use `isRTL` to determine the chevron icon direction

brunobasto committed:
More clearily separate the two contents of this modal with an early return

brunobasto committed:
Use DataForm for the confirmation form

brunobasto committed:
Fix unit tests

arthur791004 commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,192 @@
+						'Please type your username in the field below to confirm. Your account will then be gone forever.'
+					) }
+				</Text>
+				<DataForm< ConfirmationFormData >
```
Nit: We need to wrap the VStack with form and change the button type to submit so that people can submit the form via the `Enter`

brunobasto replied:
Good catch. Adding it..

brunobasto replied:
Added in https://github.com/Automattic/wp-calypso/pull/105544/commits/3ee7f9c2083851c6193d40240f926d5d31a71932

arthur791004 reviewed (approved):
Looks great now 👍

brunobasto commented:
> Looks great now 👍

Thank you for reviewing it @arthur791004 🙇

brunobasto committed:
Wrap the confirmation form with a <form> to allow submission with the `Enter` key as well

fushar commented on client/dashboard/me/profile-deletion/confirmation-modal.tsx:
```diff
@@ -0,0 +1,201 @@
+			? [
+					{
+						text: __( "Change your site's address" ),
+						to: '/v2/domains',
```
Since we're already using `RouterLinkButton`, we don't need to prepend the URL with `/v2` as it's already the base URL. In fact, we shouldn't add it, because we might want to update the prefix in the future (pgz0xU-bt-p2).

I also see that we also have `/start` link down below. It needs to be rendered with a regular `RouterLinkButton`, not a `RouterLinkButton`.

fushar commented on client/dashboard/me/profile-deletion/index.tsx:
```diff
@@ -0,0 +1,80 @@
+								isDestructive
+								variant="secondary"
+								size="compact"
+								style={ { minWidth: 'fit-content' } }
```
Do we really need this custom style?

brunobasto replied:
Good catch! Not needed. This was from previous local changes before I discovered `<ActionItem>`. I'll send a follow-up PR removing it.

fushar reviewed (commented):
In v1, I am blocked from deleting my account because I still have active purchases:

<img width="1113" height="414" alt="Image" src="https://github.com/user-attachments/assets/8370c167-8397-4512-9f11-021d7a8f8ba3" />

It seems we also need to implement this for v2?
